[Merged by Bors] - feat(measure_theory/pi_system) lemmas for pi_system, useful for independence.

mzinkevi opened the pull request:
The goal here is to prove that the expectation of a product of an finite number of independent random variables equals the production of the expectations.

See https://github.com/leanprover-community/mathlib/tree/mzinkevi_independent_finite_alt


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

mzinkevi committed:
feat(measure_theory/pi_system) lemmas for pi_system, useful for independence

mzinkevi committed:
.

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+  λ s t h_s h_t h_nonempty, generate_pi_system.inter h_s h_t h_nonempty
+
+lemma generate_pi_system_subset {α} {g t : set (set α)} (h_t : is_pi_system t)
+  (h_sub : g ⊆ t) : (generate_pi_system g : (set (set α))) ⊆ t := begin
```
```suggestion
  (h_sub : g ⊆ t) : generate_pi_system g ⊆ t :=
 begin
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+
+lemma generate_pi_system_subset {α} {g t : set (set α)} (h_t : is_pi_system t)
+  (h_sub : g ⊆ t) : (generate_pi_system g : (set (set α))) ⊆ t := begin
+  rw set.subset_def,
```
```suggestion

```
You don't need that line, intros will do it.

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+  intros x h,
+  induction h with s h_s s u h_gen_s h_gen_u h_nonempty h_s h_u,
+  apply h_sub h_s,
+  apply h_t _ _ h_s h_u h_nonempty
```
Here there are two goals, you should put them in braces.
```suggestion
  { apply h_sub h_s, },
  { apply h_t _ _ h_s h_u h_nonempty, },
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+  apply generate_pi_system_subset,
+  apply h_pi,
+  apply set.subset.refl,
+  apply generate_pi_system.base,
```
```suggestion
  generate_pi_system g = g :=
begin
  apply le_antisymm,
  { exact generate_pi_system_subset h_pi set.subset.rfl, },
  { apply generate_pi_system.base, },
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+begin
+  induction h_in_pi with s h_s s u h_gen_s h_gen_u h_nonempty h_s h_u,
+  apply h_meas_g _ h_s,
+  apply measurable_set.inter h_s h_u,
```
```suggestion
  induction h_in_pi with s h_s s u h_gen_s h_gen_u h_nonempty h_s h_u,
  { apply h_meas_g _ h_s, },
  { apply measurable_set.inter h_s h_u, },
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+
+/- This is similar to mem_generate_pi_system_Union_elim', but
+   focuses on a set of elements in b, as opposed to the whole type. -/
+lemma mem_generate_pi_system_Union_elim' {α} {β : Type*} {g : β → (set (set α))} {s: set β}
```
```suggestion
lemma mem_generate_pi_system_Union_elim' {α β} {g : β → set (set α)} {s: set β}
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+
+/- This theorem shows that every element of the pi system generated by the union of the
+   pi systems can be represented by a finite union of elements from the pi systems. -/
+lemma mem_generate_pi_system_Union_elim {α} {β : Type*} {g : β → (set (set α))}
```
```suggestion
lemma mem_generate_pi_system_Union_elim {α β} {g : β → set (set α)}
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+  apply generate_pi_system_measurable_set,
+  intros s h_s_in_g,
+  apply measurable_space.measurable_set_generate_from h_s_in_g,
+end
```
The preferred style is to put all hypotheses before the ` : `.

Since you opened `measurable_space`, you can write ` generate_from_le` instead of ` measurable_space.generate_from_le`. You should delete those `measurable_space.` in all the proofs.

The suggestion below also uses a term mode proof. Use it or keep you tactic mode proof, as you want.
```suggestion
lemma generate_from_measurable_set_of_generate_pi_system {α} {g : set (set α)} (t : set α)
  (ht : t ∈ generate_pi_system g) :
  (generate_from g).measurable_set' t :=
@generate_pi_system_measurable_set α (generate_from g) g
  (λ s h_s_in_g, measurable_set_generate_from h_s_in_g) t ht
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+    apply h_t },
+  { intros t h_t, apply measurable_space.measurable_set_generate_from,
+    apply generate_pi_system.base h_t },
+end
```
```suggestion
lemma generate_from_generate_pi_system_eq {α} {g : set (set α)} :
  generate_from (generate_pi_system g) = generate_from g :=
begin
  apply le_antisymm; apply generate_from_le,
  { exact λ t h_t, generate_from_measurable_set_of_generate_pi_system t h_t, },
  { exact λ t h_t, measurable_set_generate_from (generate_pi_system.base h_t), },
end
```
Same proof but on fewer lines, and without the `measurable_space` prefix.

RemyDegenne reviewed (commented):
Here are several minor comments, mostly about style, with suggestions to shorten the proofs.
I did not look at the two long proofs at the end yet.

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,188 @@
+    apply h_s _ h, apply h_t' _ h_1,
+    rw finset.mem_union at h_b,
+    apply false.elim (h_b.elim h h_1) },
+end
```
I made an attempt at simplifying that proof. Here is the result.
```suggestion
begin
  induction h_t with s h_s s t' h_gen_s h_gen_t' h_nonempty h_s h_t',
  { rcases h_s with ⟨t', ⟨⟨b, rfl⟩, h_s_in_t'⟩⟩,
    refine ⟨{b}, (λ _, s), _⟩,
    simpa using h_s_in_t', },
  { rcases h_t' with ⟨T_t', ⟨f_t', ⟨rfl, h_t'⟩⟩⟩,
    rcases h_s with ⟨T_s, ⟨f_s, ⟨rfl, h_s⟩ ⟩ ⟩,
    use [(T_s ∪ T_t'), (λ (b:β),
      if (b ∈ T_s) then (if (b ∈ T_t') then (f_s b ∩ (f_t' b)) else (f_s b))
      else (if (b ∈ T_t') then (f_t' b) else (∅ : set α)))],
    split,
    { ext a,
      simp_rw [set.mem_inter_iff, set.mem_Inter, finset.mem_union, or_imp_distrib],
      rw ← forall_and_distrib,
      split; intros h1 b; by_cases hbs : b ∈ T_s; by_cases hbt : b ∈ T_t'; specialize h1 b;
        simp only [hbs, hbt, if_true, if_false, true_implies_iff, and_self, false_implies_iff,
          and_true, true_and] at h1 ⊢,
      all_goals { exact h1, }, },
    intros b h_b,
    split_ifs with hbs hbt hbt,
    { refine h_pi b (f_s b) (f_t' b) (h_s b hbs) (h_t' b hbt) (set.nonempty.mono _ h_nonempty),
      exact set.inter_subset_inter (set.bInter_subset_of_mem hbs) (set.bInter_subset_of_mem hbt), },
    { exact h_s b hbs, },
    { exact h_t' b hbt, },
    { rw finset.mem_union at h_b,
      apply false.elim (h_b.elim hbs hbt), }, },
end
```

mzinkevi committed:
Apply suggestions from code review

Co-authored-by: Rémy Degenne <remydegenne@gmail.com>

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,166 @@
+
+lemma is_pi_system_generate_pi_system {α} (g : set (set α)) :
+  is_pi_system (generate_pi_system g) :=
+  λ s t h_s h_t h_nonempty, generate_pi_system.inter h_s h_t h_nonempty
```
```suggestion
λ s t h_s h_t h_nonempty, generate_pi_system.inter h_s h_t h_nonempty
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,166 @@
+
+lemma generate_pi_system_subset {α} {g t : set (set α)} (h_t : is_pi_system t)
+  (h_sub : g ⊆ t) : generate_pi_system g ⊆ t :=
+ begin
```
```suggestion
begin
```

bryangingechen replied:
This was marked resolved, but the stray space is still there.

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,166 @@
+lemma mem_generate_pi_system_Union_elim {α β} {g : β → set (set α)}
+  (h_pi : ∀ b : β, is_pi_system (g b))
+  (t : set α)
+  (h_t : t ∈ generate_pi_system (⋃ (b : β), g b)) :
```
```suggestion
  (h_pi : ∀ b, is_pi_system (g b)) (t : set α) (h_t : t ∈ generate_pi_system (⋃ b, g b)) :
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,166 @@
+lemma mem_generate_pi_system_Union_elim' {α β} {g : β → set (set α)} {s: set β}
+  (h_pi : ∀ b ∈ s, is_pi_system (g b))
+  (t : set α)
+  (h_t : t ∈ generate_pi_system (⋃ b ∈ s, g b)) :
```
```suggestion
  (h_pi : ∀ b ∈ s, is_pi_system (g b)) (t : set α) (h_t : t ∈ generate_pi_system (⋃ b ∈ s, g b)) :
```

RemyDegenne reviewed (commented):
A few minor changes about spacing.

mzinkevi committed:
Merge branch 'master' into mzinkevi_pi_system

mzinkevi committed:
Merge branch 'master' into mzinkevi_pi_system

mzinkevi committed:
Apply suggestions from code review

Co-authored-by: Rémy Degenne <remydegenne@gmail.com>

mzinkevi committed:
Merge branch 'mzinkevi_pi_system' of github.com:leanprover-community/mathlib into mzinkevi_pi_system

mzinkevi committed:
Entered note by Remy into the code

mzinkevi committed:
Reordered proof to resolve dependency

mzinkevi committed:
No longer need fin_cases: removed

mzinkevi committed:
Removed subset

mzinkevi committed:
Finally added a comment

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+  (h_pi : ∀ b ∈ s, is_pi_system (g b)) (t : set α) (h_t : t ∈ generate_pi_system (⋃ b ∈ s, g b)) :
+  (∃ (T : finset β) (f : β → set α), (↑T ⊆ s) ∧ (t = ⋂ b ∈ T, f b) ∧ (∀ b ∈ T, f b ∈ (g b))) :=
+begin
+  classical,
```
This is not needed.
```suggestion
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+  have h1 := @mem_generate_pi_system_Union_elim α (subtype s) (g ∘ subtype.val) _ t _,
+  rcases h1 with ⟨T, ⟨f,⟨ rfl, h_t'⟩ ⟩⟩,
+  use (T.image subtype.val),
+  use (function.extend subtype.val f (λ (b:β), (∅ : set α))),
```
```suggestion
  use [T.image subtype.val, function.extend subtype.val f (λ (b:β), (∅ : set α))],
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+    { simp only [set.mem_Inter, subtype.forall, finset.set_bInter_finset_image],
+      intros h1 b h_b h_b_in_T,
+      have h2 := h1 b h_b h_b_in_T, revert h2,
+      rw function.extend_apply subtype.val_injective, apply id } },
```
It is better to put only one tactic per line, unless you close the whole goal in one line, like you do at line 165.
```suggestion
      have h2 := h1 b h_b h_b_in_T,
      revert h2,
      rw function.extend_apply subtype.val_injective,
      apply id } },
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+    have h_b_alt : b = (subtype.mk b h_b_w).val := rfl,
+    rw h_b_alt,
+    rw function.extend_apply subtype.val_injective,
+    apply h_t', apply h_b_h },
```
```suggestion
    apply h_t',
    apply h_b_h },
```

mzinkevi committed:
Renamed g and used membership

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+  { have h1 : (⋃ (b : subtype s), (g ∘ subtype.val) b) = (⋃ (b : β) (H : b ∈ s), g b),
+    { ext x, simp only [exists_prop, set.mem_Union, function.comp_app, subtype.exists,
+        subtype.coe_mk], refl },
+    rw h1, apply h_t },
```
Same remark as above: put each tactic on its line. Additionally, instead of `rw h1, apply h_t` you can write `rwa h1` (it is `rw` followed by `assumption`). Finally, if you use `suffices` instead of `have`, you don't need as many braces:
```suggestion
  { suffices h1 : (⋃ (b : subtype s), (g ∘ subtype.val) b) = (⋃ b (H : b ∈ s), g b), by rwa h1,
    ext x,
    simp only [exists_prop, set.mem_Union, function.comp_app, subtype.exists, subtype.coe_mk],
    refl, },
```

RemyDegenne commented on src/measure_theory/pi_system.lean:
```diff
@@ -0,0 +1,170 @@
+    cases h_b,
+    have h_b_alt : b = (subtype.mk b h_b_w).val := rfl,
+    rw h_b_alt,
+    rw function.extend_apply subtype.val_injective,
```
```suggestion
    rw [h_b_alt, function.extend_apply subtype.val_injective],
```

RemyDegenne reviewed (commented):
I finally had a look at the last lemma.

mzinkevi committed:
Apply suggestions from code review

Thank you! These changes are helpful in understanding the style. I haven't quite figured out the style as to line breaks, but I'll try to implement this pattern moving forward.
I have seen `suffices`, but I haven't used it yet. I'll look through my next PR to see where I can use it.

Co-authored-by: Rémy Degenne <remydegenne@gmail.com>

sgouezel reviewed (commented):
Thanks a lot for this! Since you are expanding nicely on what there is on pi systems in the file `measurable_space.lean`, I wonder if it wouldn't make sense to move all there is on this notion (and on dynkin systems) to the file you are creating, to get something more self-contained and easier to locate.

By the way, you are importing `measure_space`, but I guess it would suffice to import `measurable_space` for everything to make sense (and you don't need to open `measure_theory` as there is no measure in your file).

mzinkevi committed:
Merge branch 'master' into mzinkevi_pi_system

Updating to latest version of master

mzinkevi committed:
Removing dependency on measure_space, to leave only a dependency on measurable_space.

mzinkevi committed:
Moved code from measurable_space to pi_system regarding dynkin and pi-systems

sgouezel committed:
expand docstrings, clean up
